### PR TITLE
[GAWB-1306] Orchestration: TSV parsing is broken

### PR DESCRIPTION
**Pull request by rtitle**
JIRA link: https://broadinstitute.atlassian.net/browse/GAWB-1306

I used this library for TSV parsing: https://github.com/uniVocity/univocity-parsers

Added unit tests and did some light test refactoring. Unit tests are passing.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [x] **LR** sign off
- [x] **Product Owner** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala [hunk 1]
```diff
@@ -4,187 +4,202 @@ import spray.http._
...
    */
   val empty = ""
   val onlyNewlines = "\n\n\n\n\n\n"
 
   val rowTooLong = List(
-      List("foo", "bar", "baz").mkString("\t"),
-      List("this", "line's", "fine").mkString("\t"),
-      List("this", "line's", "too", "long").mkString("\t")
-    ).mkString("\n")
+      List("foo", "bar", "baz").tabbed,
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala:30]: fancy!

File: src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala [hunk 2]
```diff
@@ -14,23 +11,31 @@ case class TSVLoadFile(
...
-      throw new RuntimeException("TSV parsing error in line " + tsvLineNo.toString + ": wrong number of fields")
-    }
-    fields.toList
+
+  private lazy val PARSER: CsvParser = {
+    // Note we're using a CsvParser with a tab delimiter rather a TsvParser.
+    // This is because the CSV formatter handles quotations correctly while the TSV formatter doesn't.
+    // See https://github.com/uniVocity/univocity-parsers#csv-format
+    val settings = new CsvParserSettings
+    settings.getFormat().setLineSeparator("\n")
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala]: Do we have to worry about `\r\n` if the TSV has come from a Windows machine?

File: src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala [hunk 3]
```diff
@@ -14,23 +11,31 @@ case class TSVLoadFile(
...
 
 object TSVParser {
-  private def parseLine( tsvLine: String, tsvLineNo: Int, nCols: Int ) = {
-    val fields = tsvLine.split("\t", -1) //note we don't trim the line here in case the last element is deliberately empty
-    if( fields.length != nCols ) {
-      throw new RuntimeException("TSV parsing error in line " + tsvLineNo.toString + ": wrong number of fields")
-    }
-    fields.toList
+
+  private lazy val PARSER: CsvParser = {
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala]: nitpick: the only place we use uppercase variables is for constants (and sometimes not even then).

File: src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala (hunk 1 above)
Comment (helgridly) [reply src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala:30]: +1 this is glorious

**Review by helgridly — CHANGES_REQUESTED** [review]
tiny nitpicks but otherwise very 👍

File: src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala (hunk 3 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala]: ok, will fix.

File: src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala (hunk 2 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala]: Hm, good question. Let me reread the docs and maybe add some test cases.

File: src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala (hunk 1 above)
Comment (rtitle) [reply src/test/scala/org/broadinstitute/dsde/firecloud/mock/MockTSV.scala:30]: implicit classes ftw

File: src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala (hunk 2 above)
Comment (rtitle) [reply src/main/scala/org/broadinstitute/dsde/firecloud/utils/TSVParser.scala]: Turns out there's a feature which seems to do what we want:
```
//You can configure the parser to automatically detect what line separator sequence is in the input
parserSettings.setLineSeparatorDetectionEnabled(true);
```

**Comment by rtitle** [comment]
@helgridly made code review comments, please take a look.

**Review by helgridly — APPROVED** [review]

**Comment by helgridly** [comment]
Winner! 👍

**Pull request merged**